### DISCUSSION TITLE
Difference between unused and timestamp filters

### DISCUSSION THREAD

**Discussion by bethatasitmay** [PaloAltoNetworks/pan-os-php#681] [Q&A]
Hi

As I understand it, the unused filter checks to see if a rule has been used/not used since a firewall was last rebooted.

And, for timestamp-last-hit.fast, it's looking at the Rule Usage data that I think is persistent across reboots.

My question is this: If the Rule Usage column in the GUI says Unused is it the same thing as the unused filter output?

I'm asking as I ran an exportToExcel with a timestamp filter "older than a year" and none of the rules with Unused in the Rule Usage column appear in the output - only rules with actual dates in the Rule Usage column older than a year appear.

**Comment by swaschkut** [comment]
both filter are done against PAN-OS XML API;

with additional argument 'debugapi' you are getting the API request you sending to the device, and also the response
copy this over into a web browser to get the response in a human readable format.

- 'filter=(rule is.unused.fast)'  against panorama DG is reaching out to all connected FW attached to this rule, to check if it is unused at all
- 'filter=(timestamp-last-hit.fast < - 365 days)'

I recommend that you are posting your filters here, you like to discuss.
I assume that your timestamp filer is wrongly defined, this is mathematical calculation

**Comment by bethatasitmay** [comment]
(timestamp-last-hit.fast < 10/01/2021)

Any rule with Rule Last Hit date in it older than 1 Oct 2021 appeared in the XLS file and I subsequently did a tag-Add.  Neither action included the rules where Last Hit was Unused.

**Reply by swaschkut** [reply]
how does the API response look like in your environment if 'Last Hit was Unused'?
which PAN-OS version are you running on your Panorama? and which version are you running on the connected Firewalls?

this framework is using Panorama only as a proxy to connect to the Firewalls to get this information from.

Maybe we should switch to Mail contact to get discuss about secure file sharing.
Please reach out to the already known contact details

**Reply by swaschkut** [reply]
maybe I understand now:


- type=rule 'filter=(timestamp-last-hit.fast < 10/01/2021)'
are actual also showing rules where timestamp-last-hit is 0 (which is unused)

- type=rule 'filter=(timestamp-last-hit.fast = 0)'
is actual a bug and error out

both scenarios are fixed/improved in the developer version

would be good if you can test this:

developer version:
https://github.com/swaschkut/pan-os-php
or
docker run --name panosphp --rm -v ${PWD}:/share -it swaschkut/pan-os-php:developer

**Reply by swaschkut** [reply]
the actual developer feature behaviour is:

**'filter=(rule is.unused.fast)'** is identical to **'filter=(timestamp-last.hit.fast = 0)'**

**type=rule 'filter=(timestamp-last-hit.fast < 10/01/2021)'** is now only showing rules which had a hit since this date

**Reply by bethatasitmay** [reply]
Well the 2nd was working before as you describe - it did not include Unused.  But, now I understand to use both filters in the future.

I'll test it out later today or tomorrow (UTC-8 timezone).

**Comment by bethatasitmay** [comment]
Sorry, I haven't had time to look at this.  Is it still in the dev build only?

**Reply by swaschkut** [reply]
I should publish v2.0.65 which include the fix/improvements from above
